u_roller.go

```diff
@@ -0,0 +1,110 @@
+package tls
+
+import (
+	"crypto/rand"
+	"encoding/binary"
+	"net"
+	"sync"
+	"time"
+)
+
+type Roller struct {
+	HelloIDs            []ClientHelloID
+	HelloIDMu           sync.Mutex
+	NextHelloID         ClientHelloID
+	TcpDialTimeout      time.Duration
+	TlsHandshakeTimeout time.Duration
+}
+
+// NewRoller creates Roller object with default range of HelloIDs to cycle through until a
+// working/unblocked one is found.
+func NewRoller() (*Roller, error) {
+	tcpDialTimeoutInc, err := getRandUint16()
+	if err != nil {
+		return nil, err
+	}
+	tcpDialTimeoutInc = tcpDialTimeoutInc % 14
+
+	tlsHandshakeTimeoutInc, err := getRandUint16()
+	if err != nil {
+		return nil, err
+	}
+	tlsHandshakeTimeoutInc = tlsHandshakeTimeoutInc % 20
+
+	r := &Roller{
+		HelloIDs: []ClientHelloID{
+			HelloChrome_Auto,
+			HelloFirefox_Auto,
+			HelloIOS_Auto,
+		},
+		TcpDialTimeout:      time.Second * time.Duration(7+int(tcpDialTimeoutInc)),
+		TlsHandshakeTimeout: time.Second * time.Duration(11+int(tlsHandshakeTimeoutInc)),
+	}
+
+	// choose random clientHelloID to try next
```

Comment:
Can you just shuffle `HelloIDs` instead?  That would be clearer.

Reply:
Not sure I agree. Instead of choosing random index, if I randomize the array I'd be able to always pick 0th element in it -- sounds like ~same level of clarity. Logic for "finding which element to evict from array" will have to remain, since last successful ID could be anywhere in shuffled array.

Reply:
I don't know what the intended behavior of the roller is, because you haven't documented it.  However, I don't think the exact behavior is important, so it makes sense to choose the behavior that is clearest to implement.

I don't think there's any need to "evict elements from an array".  Instead, you could just check
1. Do I have a known working hello ID?  If so, use it.
2. Otherwise, try each entry in the list.
3. If one worked, record it as a known working ID.

If you don't like that heuristic, then there are plenty of other simple options like "If it worked, move it to the front of the list.".  The current heuristic seems unnecessarily complex.